third_party/move/move-compiler/src/typing/expand.rs

```diff
@@ -57,21 +57,65 @@ fn types(context: &mut Context, ss: &mut Vec<Type>) {
     }
 }
 
-pub fn type_(context: &mut Context, ty: &mut Type) {
+// type_ for expanding the `i`-th type param of struct `ty_name`
+fn type_struct_ty_param(context: &mut Context, ty: &mut Type, i: usize, ty_name: &TypeName_) {
+    match ty_name {
+        TypeName_::ModuleType(mod_id, struct_name) => {
+            let param_name = &context
+                .struct_definition(mod_id, struct_name)
+                .type_parameters[i]
+                .param
+                .user_specified_name
+                .value;
+            let msg = format!("Cannot infer the type parameter `{param_name}` for generic struct `{ty_name}`. Try providing a type parameter.");
+            type_msg_(context, ty, &msg);
+        },
+        _ => type_(context, ty),
+    }
+}
+
+fn type_fun_ty_param(
+    context: &mut Context,
+    ty: &mut Type,
+    i: usize,
+    m: &ModuleIdent,
+    n: &FunctionName,
+) {
+    let param_name = context.function_info(m, n).signature.type_parameters[i].user_specified_name;
+    type_msg_(context, ty, &format!("Cannot infer the type parameter `{param_name}` for generic function `{m}::{n}`. Try providing a type parameter."));
+}
+
+// Try to expand the type of ty, warning with msg_uninferred if type cannot be inferred.
+fn type_msg_(context: &mut Context, ty: &mut Type, msg_uninferred: &str) {
     use Type_::*;
     match &mut ty.value {
         Anything | UnresolvedError | Param(_) | Unit => (),
         Ref(_, b) => type_(context, b),
         Var(tvar) => {
             let ty_tvar = sp(ty.loc, Var(*tvar));
-            let replacement = core::unfold_type(&context.subst, ty_tvar);
+            fn unfold_type_mut(subst: &mut Subst, sp!(loc, t_): Type) -> Type {
+                match t_ {
+                    Type_::Var(i) => {
+                        let last_tvar = forward_tvar(subst, i);
+                        match subst.get(last_tvar) {
+                            Some(sp!(_, Type_::Var(_))) => unreachable!(),
+                            None => {
+                                subst.insert(last_tvar, sp(loc, Type_::UnresolvedError));
+                                sp(loc, Type_::Anything)
+                            },
+                            Some(inner) => inner.clone(),
+                        }
+                    },
+                    x => sp(loc, x),
+                }
+            }
+            let replacement = unfold_type_mut(&mut context.subst, ty_tvar);
             let replacement = match replacement {
                 sp!(_, Var(_)) => panic!("ICE unfold_type_base failed to expand"),
                 sp!(loc, Anything) => {
-                    let msg = "Could not infer this type. Try adding an annotation";
                     context
```

Comment:
This is to avoid duplicate error messages for uninferred type variables. In the first time they are resolved to Err(_), and Anything for all following queries.

Reply:
Please add a comment with this text.